File: packages/ui-android/chorus-ui/src/main/kotlin/dev/blinddsai/chorus/ui/button/ChorusButton.kt
package dev.blinddsai.chorus.ui.button

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.ripple.rememberRipple
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.painter.Painter
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.foundation.Image
import androidx.compose.material.Text
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import dev.blinddsai.chorus.tokens.ChorusTheme
import dev.blinddsai.chorus.tokens.generated.ChorusBorderWidth
import dev.blinddsai.chorus.tokens.generated.ChorusIconSize
import dev.blinddsai.chorus.tokens.generated.ChorusLayout
import dev.blinddsai.chorus.tokens.generated.ChorusRadius
import dev.blinddsai.chorus.tokens.generated.ChorusStateOpacity
import dev.blinddsai.chorus.tokens.generated.ChorusTypography

/**
 * Compose port of the Chorus standard Button.
 *
 * Spec of record: `schema/components/button/standard.spec.json`. Two
 * independent axes — [size] (chosen by surface) and [appearance] (chosen
 * by emphasis) — plus the [flavor] axis ([Flavor.Destructive] swaps the
 * primary family for the error family). Pointer-state overlays
 * (pressed) compose foreground-over-base per the spec's "Compose state
 * as foreground-over-base" rule; disabled lowers the container opacity
 * instead.
 */
enum class ChorusButtonSize { Large, Medium, Small }
enum class ChorusButtonAppearance { Primary, Secondary, Outlined, Tertiary }
enum class ChorusButtonFlavor { Standard, Destructive }

@Composable
fun ChorusButton(
    label: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    size: ChorusButtonSize = ChorusButtonSize.Large,
    appearance: ChorusButtonAppearance = ChorusButtonAppearance.Primary,
    flavor: ChorusButtonFlavor = ChorusButtonFlavor.Standard,
    leadingIcon: Painter? = null,
    fullWidth: Boolean = false,
    truncate: Boolean = false,
    enabled: Boolean = true,
) {
    val s = sizing(size)
    val surface = surfaceFor(appearance, flavor)

    val interactionSource = remember { MutableInteractionSource() }
    val pressed by interactionSource.collectIsPressedAsState()

    val shape = RoundedCornerShape(s.radius)

    Box(
        modifier = modifier
            .then(if (fullWidth) Modifier.fillMaxWidth() else Modifier.widthIn(min = 160.dp))
            .defaultMinSize(minHeight = s.minHeight)
            .alpha(if (enabled) 1f else ChorusStateOpacity.disabled)
            .clip(shape)
            .background(surface.background, shape)
            .then(
                surface.border?.let { (w, c) ->
                    Modifier.border(BorderStroke(w, c), shape)
                } ?: Modifier
            )
            .clickable(
                interactionSource = interactionSource,
                indication = rememberRipple(color = surface.label),
                enabled = enabled,
                role = Role.Button,
                onClick = onClick,
            )
            .padding(PaddingValues(horizontal = s.paddingInline, vertical = s.paddingBlock)),
        contentAlignment = Alignment.Center,
    ) {
        // Pressed overlay — foreground-color tinted scrim on top of base.
        if (pressed) {
            Box(
                modifier = Modifier
                    .matchParentSize()
                    .background(surface.label.copy(alpha = ChorusStateOpacity.pressed))
            )
        }

        Row(
            horizontalArrangement = Arrangement.spacedBy(s.gap, Alignment.CenterHorizontally),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            if (leadingIcon != null) {
                Image(
                    painter = leadingIcon,
                    contentDescription = null,
                    modifier = Modifier.size(s.iconSize),
                )
            }
            Text(
                text = label,
                style = s.labelTypo,
                color = surface.label,
                maxLines = 1,
                overflow = if (truncate) TextOverflow.Ellipsis else TextOverflow.Clip,
            )
        }
    }
}

private data class Sizing(
    val paddingBlock: Dp,
    val paddingInline: Dp,
    val gap: Dp,
    val minHeight: Dp,
    val radius: Dp,
    val iconSize: Dp,
    val labelTypo: TextStyle,
)

private fun sizing(size: ChorusButtonSize): Sizing = when (size) {
    ChorusButtonSize.Large -> Sizing(
        paddingBlock = ChorusLayout.Container.xs,
        paddingInline = ChorusLayout.Container.md,
        gap = ChorusLayout.Inline.md,
        minHeight = 48.dp, // ref.space.600
        radius = ChorusRadius.md,
        iconSize = ChorusIconSize.lg,
        labelTypo = ChorusTypography.Label.lg,
    )
    ChorusButtonSize.Medium -> Sizing(
        paddingBlock = ChorusLayout.Container.xs,
        paddingInline = ChorusLayout.Container.md,
        gap = ChorusLayout.Inline.md,
        minHeight = 40.dp, // ref.space.500
        radius = ChorusRadius.md,
        iconSize = ChorusIconSize.md,
        labelTypo = ChorusTypography.Label.md,
    )
    ChorusButtonSize.Small -> Sizing(
        paddingBlock = ChorusLayout.Container.s2xs,
        paddingInline = ChorusLayout.Container.sm,
        gap = ChorusLayout.Inline.sm,
        minHeight = 32.dp, // ref.space.400
        radius = ChorusRadius.sm,
        iconSize = ChorusIconSize.md,
        labelTypo = ChorusTypography.Label.md,
    )
}

private data class Surface(
    val background: Color,
    val label: Color,
    val border: Pair<Dp, Color>?,
)

@Composable
private fun surfaceFor(
    appearance: ChorusButtonAppearance,
    flavor: ChorusButtonFlavor,
): Surface {
    val c = ChorusTheme.colors
    return when (flavor to appearance) {
        ChorusButtonFlavor.Standard to ChorusButtonAppearance.Primary ->
            Surface(c.primary, c.onPrimary, null)
        ChorusButtonFlavor.Standard to ChorusButtonAppearance.Secondary ->
            Surface(c.secondaryContainer, c.onSecondaryContainer, null)
        ChorusButtonFlavor.Standard to ChorusButtonAppearance.Outlined ->
            Surface(Color.Transparent, c.primary, ChorusBorderWidth.hairline to c.primary)
        ChorusButtonFlavor.Standard to ChorusButtonAppearance.Tertiary ->
            Surface(Color.Transparent, c.onSurfaceVariant, null)
        ChorusButtonFlavor.Destructive to ChorusButtonAppearance.Primary ->
            Surface(c.error, c.onError, null)
        ChorusButtonFlavor.Destructive to ChorusButtonAppearance.Secondary ->
            Surface(c.errorContainer, c.onErrorContainer, null)
        ChorusButtonFlavor.Destructive to ChorusButtonAppearance.Outlined ->
            Surface(Color.Transparent, c.error, ChorusBorderWidth.hairline to c.error)
        ChorusButtonFlavor.Destructive to ChorusButtonAppearance.Tertiary ->
            Surface(Color.Transparent, c.error, null)
        else -> error("unreachable")
    }
}

File: packages/ui-ios/Sources/ChorusUI/Button/ChorusButton.swift
import SwiftUI
import ChorusTokens

/// SwiftUI port of the Chorus standard Button.
///
/// Spec of record: `schema/components/button/standard.spec.json`.
/// Two independent axes — `size` (chosen by surface) and `appearance`
/// (chosen by emphasis) — plus a `destructive` flavor that swaps the
/// primary family for the error family. Pointer-state overlays
/// (hover / pressed / focused) are composed as a foreground-over-base
/// label-colored layer per the spec's "Compose state as foreground-over-base"
/// rule. Disabled is the only state that lowers container opacity rather
/// than overlaying.

public struct ChorusButton: View {
    public enum Size { case large, medium, small }
    public enum Appearance { case primary, secondary, outlined, tertiary }
    public enum Flavor { case standard, destructive }

    private let label: String
    private let size: Size
    private let appearance: Appearance
    private let flavor: Flavor
    private let leadingIcon: Image?
    private let fullWidth: Bool
    private let truncate: Bool
    private let action: () -> Void

    @Environment(\.colorScheme) private var colorScheme
    @Environment(\.isEnabled) private var isEnabled

    public init(
        _ label: String,
        size: Size = .large,
        appearance: Appearance = .primary,
        flavor: Flavor = .standard,
        leadingIcon: Image? = nil,
        fullWidth: Bool = false,
        truncate: Bool = false,
        action: @escaping () -> Void
    ) {
        self.label = label
        self.size = size
        self.appearance = appearance
        self.flavor = flavor
        self.leadingIcon = leadingIcon
        self.fullWidth = fullWidth
        self.truncate = truncate
        self.action = action
    }

    public var body: some View {
        let s = sizing(size)
        let palette = ChorusColors.themed(for: colorScheme)
        let surface = surfaceFor(palette: palette)

        Button(action: action) {
            HStack(spacing: s.gap) {
                if let icon = leadingIcon {
                    icon
                        .resizable()
                        .renderingMode(.template)
                        .frame(width: s.iconSize, height: s.iconSize)
                }
                Text(label)
                    .lineLimit(1)
                    .truncationMode(truncate ? .tail : .tail)
                    .fixedSize(horizontal: !truncate && !fullWidth, vertical: false)
            }
            .frame(maxWidth: fullWidth ? .infinity : nil, minHeight: s.minHeight)
            .frame(minWidth: fullWidth ? nil : 160)
            .padding(.vertical, s.paddingBlock)
            .padding(.horizontal, s.paddingInline)
        }
        .buttonStyle(ChorusButtonStyle(surface: surface, radius: s.radius, label: s.label))
        .font(.system(size: s.labelTypo.size, weight: s.labelTypo.fontWeight))
        .tracking(s.labelTypo.trackingPoints)
        .opacity(isEnabled ? 1.0 : ChorusStateOpacity.disabled)
    }

    // MARK: - Sizing -------------------------------------------------------
    private struct Sizing {
        let paddingBlock: CGFloat
        let paddingInline: CGFloat
        let gap: CGFloat
        let minHeight: CGFloat
        let radius: CGFloat
        let labelTypo: ChorusTextStyle
        let iconSize: CGFloat
        let label: ChorusTextStyle
    }

    private func sizing(_ size: Size) -> Sizing {
        switch size {
        case .large:
            return Sizing(
                paddingBlock: ChorusLayout.Container.xs,
                paddingInline: ChorusLayout.Container.md,
                gap: ChorusLayout.Inline.md,
                minHeight: 48, // ref.space.600
                radius: ChorusRadius.md,
                labelTypo: ChorusTypography.Label.lg,
                iconSize: ChorusIconSize.lg,
                label: ChorusTypography.Label.lg
            )
        case .medium:
            return Sizing(
                paddingBlock: ChorusLayout.Container.xs,
                paddingInline: ChorusLayout.Container.md,
                gap: ChorusLayout.Inline.md,
                minHeight: 40, // ref.space.500
                radius: ChorusRadius.md,
                labelTypo: ChorusTypography.Label.md,
                iconSize: ChorusIconSize.md,
                label: ChorusTypography.Label.md
            )
        case .small:
            return Sizing(
                paddingBlock: ChorusLayout.Container.s2xs,
                paddingInline: ChorusLayout.Container.sm,
                gap: ChorusLayout.Inline.sm,
                minHeight: 32, // ref.space.400
                radius: ChorusRadius.sm,
                labelTypo: ChorusTypography.Label.md,
                iconSize: ChorusIconSize.md,
                label: ChorusTypography.Label.md
            )
        }
    }

    // MARK: - Appearance ---------------------------------------------------
    struct Surface {
        let background: Color
        let label: Color
        let border: (width: CGFloat, color: Color)?
    }

    private func surfaceFor(palette: ChorusThemedColors) -> Surface {
        switch (flavor, appearance) {
        case (.standard, .primary):
            return Surface(background: palette.primary, label: palette.onPrimary, border: nil)
        case (.standard, .secondary):
            return Surface(background: palette.secondaryContainer, label: palette.onSecondaryContainer, border: nil)
        case (.standard, .outlined):
            return Surface(background: .clear, label: palette.primary, border: (ChorusBorderWidth.hairline, palette.primary))
        case (.standard, .tertiary):
            return Surface(background: .clear, label: palette.onSurfaceVariant, border: nil)
        case (.destructive, .primary):
            return Surface(background: palette.error, label: palette.onError, border: nil)
        case (.destructive, .secondary):
            return Surface(background: palette.errorContainer, label: palette.onErrorContainer, border: nil)
        case (.destructive, .outlined):
            return Surface(background: .clear, label: palette.error, border: (ChorusBorderWidth.hairline, palette.error))
        case (.destructive, .tertiary):
            return Surface(background: .clear, label: palette.error, border: nil)
        }
    }
}

// MARK: - ButtonStyle ------------------------------------------------------

/// Renders the background, border, label color, and pointer-state overlays.
/// Disabled-state container dim is applied by the parent view's `.opacity`
/// modifier so this style only handles pressed (and would handle hovered
/// on macOS if needed).
private struct ChorusButtonStyle: ButtonStyle {
    let surface: ChorusButton.Surface
    let radius: CGFloat
    let label: ChorusTextStyle

    func makeBody(configuration: Configuration) -> some View {
        ZStack {
            // Base background.
            RoundedRectangle(cornerRadius: radius, style: .continuous)
                .fill(surface.background)

            // Foreground-over-base state overlay (pressed).
            if configuration.isPressed {
                RoundedRectangle(cornerRadius: radius, style: .continuous)
                    .fill(surface.label.opacity(ChorusStateOpacity.pressed))
            }

            // Outline.
            if let border = surface.border {
                RoundedRectangle(cornerRadius: radius, style: .continuous)
                    .strokeBorder(border.color, lineWidth: border.width)
            }

            configuration.label
                .foregroundStyle(surface.label)
        }
        .compositingGroup()
    }
}
